### PR TITLE
Structured names

### PR DESCRIPTION
**Pull request by robdockins**
Update to track API changes in saw-core, without incorporating the resolve value updates from #887.  This should help us to pin down the source of the proof slowdowns seen in that PR.

### CODE REVIEW

**Comment by robdockins** [comment]
CI runs confirm that the slowdown in #887 is due to the changes in the setup strategy, not the other API updates.

We should merge this in the meantime to keep up with upstream submodules and avoid potentially holding up other in-flight PRs.

File: src/SAWScript/JavaBuiltins.hs [hunk 1]
```diff
@@ -327,7 +329,7 @@ mkMixedExpr t = do
   sc <- lift getSharedContext
   let exts = getAllExts t
       extNames = map ecName exts
-  jes <- mapM (getJavaExpr "mkMixedExpr") extNames
+  jes <- mapM (getJavaExpr "mkMixedExpr" . Text.unpack . toShortName) extNames -- TODO?????
```
Comment (andreistefanescu) [inline src/SAWScript/JavaBuiltins.hs:332]: what's the TODO?

File: src/SAWScript/JavaExpr.hs [hunk 2]
```diff
@@ -164,7 +166,7 @@ jeVarName = map dotToUnderscore . ppJavaExpr
         dotToUnderscore c = c
 
 asJavaExpr :: Term -> Maybe String
-asJavaExpr (asExtCns -> Just ec) = Just (ecName ec)
+asJavaExpr (asExtCns -> Just ec) = Just (Text.unpack (toShortName (ecName ec))) -- TODO?
```
Comment (andreistefanescu) [inline src/SAWScript/JavaExpr.hs:169]: what's the TODO?

File: src/SAWScript/Builtins.hs [hunk 3]
```diff
@@ -468,12 +474,49 @@ printGoalSize =
...
+--  in the SAWCore naming environment.  If both stages
+--  fail, then throw an exception.
+resolveName :: SharedContext -> String -> TopLevel VarIndex
+resolveName sc nm =
+  do cenv <- rwCryptol <$> getTopLevelRW
+     let ?fileReader = StrictBS.readFile
+     res <- io $ CEnv.resolveIdentifier cenv tnm
+     case res of
+       Just cnm ->
+         do x <- io $ Cryptol.importName cnm
```
Comment (andreistefanescu) [inline src/SAWScript/Builtins.hs:498]: let's avoid using names such as `x` and `y` here

File: src/SAWScript/Builtins.hs [hunk 4]
```diff
@@ -468,12 +474,49 @@ printGoalSize =
...
+--
+--   We first attempt to find the name in the local Cryptol
+--   environment; if the name is found, attempt to resolve it to
+--   an 'ExtCns' in the SAWCore environment.  If the given name
+--   does not resolve to a cryptol value in the current environment that
+--  maps to an 'ExtCns', then instead directly look it up
+--  in the SAWCore naming environment.  If both stages
+--  fail, then throw an exception.
+resolveName :: SharedContext -> String -> TopLevel VarIndex
+resolveName sc nm =
```
Comment (andreistefanescu) [inline src/SAWScript/Builtins.hs:492]: how does the error message look like when the string is not a valid name?

File: src/SAWScript/Crucible/LLVM/Builtins.hs [hunk 5]
```diff
@@ -304,6 +304,22 @@ crucible_llvm_array_size_profile assume bic opts (Some lm) nm lemmas setup = do
     profiles <- io $ readIORef cell
     pure . fmap (\(fnm, prof) -> (Text.unpack fnm, prof)) $ Map.toList profiles
 
+llvmURI :: String -> URI
+llvmURI symbol_name =
+  fromMaybe (error $ unwords ["mkLLVMName", "Could not create LLVM symbol name", symbol_name]) $
```
Comment (brianhuffman) [inline src/SAWScript/Crucible/LLVM/Builtins.hs:309]: Is this call to `error`believed to be unreachable? Assuming so, it probably ought to be a call to `panic`; are we using `panic` within saw-script yet?

**Review by brianhuffman — APPROVED** [review]
It all looks reasonable to me.

File: src/SAWScript/JavaBuiltins.hs (hunk 1 above)
Comment (robdockins) [reply src/SAWScript/JavaBuiltins.hs:332]: The TODO here is that using `toShortName` for anything other than user output is very suspicious.  This line is no worse than what was there before (I think), but someone who understands how the Java stuff is put together should think about what happens if there are name conflicts and such.

File: src/SAWScript/JavaExpr.hs (hunk 2 above)
Comment (robdockins) [reply src/SAWScript/JavaExpr.hs:169]: Same as above, `toShortName` can have name collisions.

File: src/SAWScript/Crucible/LLVM/Builtins.hs (hunk 5 above)
Comment (robdockins) [reply src/SAWScript/Crucible/LLVM/Builtins.hs:309]: It looks like `panic` hasn't migrated up to SAWScript yet.  I think this error should be impossible, but I'll have to figure out if there are conditions under which `mkPathPiece` might fail.

File: src/SAWScript/Builtins.hs [hunk 6]
```diff
@@ -775,40 +821,58 @@ proveUnintYices = proveUnintSBV SBV.yices
...
 
 w4_yices :: ProofScript SV.SatResult
-w4_yices = wrapW4Prover $ Prover.proveWhat4_yices []
+w4_yices = wrapW4Prover $ Prover.proveWhat4_yices mempty
 
 w4_unint_boolector :: [String] -> ProofScript SV.SatResult
-w4_unint_boolector = wrapW4Prover . Prover.proveWhat4_boolector
+w4_unint_boolector unints =
+  do sc <- lift $ SV.getSharedContext
+     unintSet <- lift $ resolveNames sc unints
```
[the remark this answers is not in the document]
Comment (robdockins) [reply src/SAWScript/Builtins.hs:838]: Maybe.  I'm pretty skeptical about the value of these wrapping combinators, but we can use them if you'd prefer.

File: src/SAWScript/JavaBuiltins.hs (hunk 1 above)
Comment (brianhuffman) [reply src/SAWScript/JavaBuiltins.hs:332]: We shouldn't worry about this code too much; it's only used by the deprecated jss-based java verification stuff. We will be removing it before too long.

File: src/SAWScript/Builtins.hs (hunk 4 above)
Comment (robdockins) [reply src/SAWScript/Builtins.hs:492]: You can see them here https://github.com/GaloisInc/saw-core/blob/94c8e51c509cb3379fec3a8710c442b8a2b6ae58/saw-core/src/Verifier/SAW/SharedTerm.hs#L408

You'll get a message that looks like:

```
Could not resolve name: "asdf"
```

or 

```
Ambiguous name "asdf" might refer to any of the following:
Module1::asdf
Module2::asdf
```

Comment (andreistefanescu) [reply src/SAWScript/Builtins.hs:492]: I mean, how does the whole error message look like, as printed by saw. Does it contain location information?

Comment (robdockins) [reply src/SAWScript/Builtins.hs:492]:
```
$ cat test.saw

unfold_term ["asdf"] {{ 42 }};

$ saw test.saw

[23:14:23.327] Loading file "/Users/rdockins/code/saw-script/test.saw"
[23:14:23.343] Stack trace:
"unfold_term" (/Users/rdockins/code/saw-script/test.saw:2:1-2:12):
Could not resolve name: "asdf"
```

Comment (andreistefanescu) [reply src/SAWScript/Builtins.hs:494]: that's what I was looking for, thanks!

**Review by andreistefanescu — APPROVED** [review]
Looks good, thanks @robdockins

**Comment by brianhuffman** [comment]
After #922, the other crucible-jvm-related changes I have lined up will require coordinated PRs on the crucible repo, which already has the structured-name stuff merged in. Basically, this means that I'll need this PR merged into master on saw-script before I can make any more crucible-jvm PRs.

**Comment by robdockins** [comment]
At this point, I'm just waiting on green lights from CI

**Comment by robdockins** [comment]
@jared-w , I'm noticing that the SAW s2n tests stall and wait for _all_ of the builds to complete before starting, even though I think they only require the linux builds.  As I'm looking at this, the linux and windows builds are all finished, and one of the OSX builds hasn't even started yet, so this adds nontrivial delay.

Would it be possible to loosen the dependencies somehow so the s2n tests could start as soon as the actual build they rely on is done?

**Pull request merged**